posts/2022-07-12-components-with-dynamic-attributes.md

```diff
@@ -0,0 +1,272 @@
+
+This function is Useful for transforming caller assigns into dynamic attributes while stripping reserved keys from the result.
+
+Now that we are assured we can transform the assigns passed to card/1 into attributes, let’s go ahead and add class assign and pass bg-* i.e “bg-blue” as the value. “bg-*” represents background color with a css added property.
+
+
+```elixir
+<h3>Work</h3>
+<%= for card <- @work_cards do %> 
+   <.card card={card} class={"bg-blue"}/>
+ <% end %>
+.
+.
+.
+ <h3>House</h3>
+ <%= for card <- @house_cards do %> 
+   <.card card={card} class={"bg-green"}/>
+ <% end %>
+.
+.
+.
+<h3>School</h3>
+<%= for card <- @school_cards do %> 
+  <.card card={card} class={"bg-yellow"}/>
+<% end %>
+
+```
+
+Inspect the assigns and see what it holds:
+
+```elixir
+def card(assigns) do
+IO.inspect(assigns, label: "==================card component=====")    
+
+~H"""
+    <div>
+      <div class="column">
+        <%= @card.task %>
+      </div>
+    </div>
+    """
+  end
+
+```
+
+```elixir
+ output:
+==================card component=====: %{
+  __changed__: nil,
+  card: %{id: "1-work", task: "deploy to production"},
+  class: "bg-blue"
+}
+```
+
+We can see the assigns contains the card and the class assigns. Lets go ahead and invoke the assigns_to_attribute/2 inside the card/1 to transform our class assign for use in `<div>` tag attribute.
+
+```elixir
+def card(assigns) do
+extra = assigns_to_attributes(assigns)
+.
+. 
+end
+```
+
+If we invoke assigns_to_attributes/2 with the assigns, it returns a list of keywords as shown:
+
+```elixir
+[
+  card: %{id: "3-school", task: "work on school project"},
+  class: "column bg-yellow"
+]
+```
+
+We don’t want to use the card assign as an attribute so we will have to exclude it from the list and only remain with the class assign.
+
+```elixir
+def card(assigns) do
+extra = assigns_to_attributes(assigns, [:card])
+assigns = assign(assigns, :extra, extra)~H"""
+    <div>
+      <div {@ extra} >
+        <%= @card.task %>
+      </div>
+    </div>
+    """
+end
+```
+
+Next, I have updated our assigns with the attributes(extra) using the [assign/2](https://hexdocs.pm/phoenix_live_view/0.16.0/Phoenix.LiveView.html#assign/2) function .We shall use the @extra to output as HTML attributes on the <div> tag.
+
+**NB:** The component markup has no column class passed to it.I had to remove it there and pass it to class assign when calling the card/1 function.
+
+```elixir
+<.card card={card} class={"column bg-green"}/>
+```
+
+This is how our rendered markup looks like when you inspect on the browser:
+
+```
+<div>
+  <div class="column bg-green">
+    wash my dog
+  </div>
+</div>
+```
+
+![board plan output2](/images/board-plan-output2.png)
```

Comment:
Wrap up with a short "Conclusion" section that sums up what the reader learned in a few sentences.